hazelcast/include/hazelcast/client/exception/protocol_exceptions.h

```diff
@@ -319,12 +325,38 @@ DEFINE_RETRYABLE_EXCEPTION_CLASS(wrong_target, protocol::WRONG_TARGET);
 DEFINE_RETRYABLE_EXCEPTION_CLASS(target_not_replica,
                                  protocol::TARGET_NOT_REPLICA_EXCEPTION);
 
+class query : public hazelcast_
+{
+public:
+    explicit query(std::string source = "",
+                   std::string message = "",
+                   std::string details = "",
+                   std::exception_ptr cause = nullptr);
+
+    query(int32_t code,
+          std::string message,
+          std::exception_ptr cause = nullptr,
+          boost::uuids::uuid originating_member_id = boost::uuids::nil_uuid(),
+          std::string suggestion = "");
+
+    int32_t code() const;
```

Comment:
In the project, coding guideline is that do not put any implementation in header unless it is needed (as required by templates).

Reply:
In the project, coding guideline is that do not put any implementation in header unless it is needed (as required by templates).

Reply:
What's the reason for that? Asking just to learn.

Reply:
I think having these small functions to be inlined would be a plus, but I guess you could just use `inline` with them.

Reply:
`inline` is a hint to compiler. It is not guaranteed by compiler. When we do the release build the compiler does optimization and may inline a lot of things that we do not inline, hence even if the code is in the cpp file and the compiler thinks that it can be optimized, it is going to inline it anyway. Therefore, I leave such optimizations to compiler to handle unless that is a critical highly used path.